sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,687 @@
+        Notes in k_init for more details.
+
+        'random': choose `n_clusters` observations (rows) at random from data
+        for the initial centroids.
+
+        If an array is passed, it should be of shape (n_clusters, n_features)
+        and gives the initial centers.
+
+        If a callable is passed, it should take arguments X, n_clusters and a
+        random state and return an initialization.
+
+    n_init : int, default=10
+        Number of time the k-means algorithm will be run with different
+        centroid seeds in each bisection.
+        That will result producing for each bisection best output of n_init
+        consecutive runs in terms of inertia.
+
+    random_state : int, RandomState instance or None, default=None
+        Determines random number generation for centroid initialization. Use
+        an int to make the randomness deterministic.
+        See :term:`Glossary <random_state>`.
+
+    max_iter : int, default=300
+        Maximum number of iterations of the k-means algorithm for a
+        single run.
+
+    verbose : int, default=0
+        Verbosity mode.
+
+    tol : float, default=1e-4
+        Relative tolerance with regards to Frobenius norm of the difference
+        in the cluster centers of two consecutive iterations to declare
+        convergence.
+
+    copy_x : bool, default=True
+        When pre-computing distances it is more numerically accurate to center
+        the data first. If copy_x is True (default), then the original data is
+        not modified. If False, the original data is modified, and put back
+        before the function returns, but small numerical differences may be
+        introduced by subtracting and then adding the data mean. Note that if
+        the original data is not C-contiguous, a copy will be made even if
+        copy_x is False. If the original data is sparse, but not in CSR format,
+        a copy will be made even if copy_x is False.
+
+    algorithm : {"auto", "full", "elkan"}, default="auto"
+        K-means algorithm to use. The classical EM-style algorithm is "full".
+        The "elkan" variation is more efficient on data with well-defined
+        clusters, by using the triangle inequality. However it's more memory
+        intensive due to the allocation of an extra array of shape
+        (n_samples, n_clusters).
+        For now "auto" (kept for backward compatibiliy) chooses "elkan" but it
+        might change in the future for a better heuristic.
+
+    bisect_strategy : {"biggest_sse", "child_biggest_sse", "largest_cluster"},
+        default="biggest_sse"
+        Defines how should bisection by performed.
+        - "biggest_sse" means that Bisect K-Means will always check
+        all calculated cluster for cluster with biggest SSE
+        (Sum of squared errors) and bisect it. That way calculated clusters
+        will be more balanced
+        - "child_biggest_sse" means that Bisect K-Means will always check
```

Comment:
I personally doubt this strategy is very useful, because the clusters will be extremely unbalanced. Can you find an example where it would be useful ?

Reply:
This is still unclear to me, but if we factorize the code for the three strategies, the cost of having this additional strategy will be low.

Reply:
I've split factorization into two commits, so it might be quite easier to remove, in case you will consider that to be thrown away